ctpolicy/ctpolicy.go

```diff
@@ -49,13 +53,135 @@ func (group *LogGroupInfo) setMinInclusions(i int) error {
 		if included(&l) {
 			group.LogURLs[l.URL] = true
+			group.LogWeights[l.URL] = 1.0
 		}
 	}
 }
 
+// satisfyMinimalInclusion returns whether number of positive weights is
+// bigger or equal to minimal inclusion number.
+func (group *LogGroupInfo) satisfyMinimalInclusion(weights map[string]float32) bool {
+	nonZeroNum := 0
+	for logURL, w := range weights {
+		if group.LogURLs[logURL] && w > 0.0 {
+			nonZeroNum++
+			if nonZeroNum >= group.MinInclusions {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// SetLogWeights applies suggested weights to the Log-group. Does not reset
+// weights and returns error when there are not enough positive weights
+// provided to reach minimal inclusion number.
+func (group *LogGroupInfo) SetLogWeights(weights map[string]float32) error {
+	for logURL, w := range weights {
+		if w < 0.0 {
+			return fmt.Errorf("trying to assign negative weight %v to Log %q", w, logURL)
+		}
+	}
+	if !group.satisfyMinimalInclusion(weights) {
+		return fmt.Errorf("trying to assign weights %v resulting into unability to reach minimal inclusion number %d", weights, group.MinInclusions)
+	}
+	group.wMu.Lock()
+	defer group.wMu.Unlock()
+	// All group weights initially reset to 0.0
+	for logURL := range group.LogURLs {
+		group.LogWeights[logURL] = 0.0
+	}
+	for logURL, w := range weights {
+		if group.LogURLs[logURL] {
+			group.LogWeights[logURL] = w
+		}
+	}
+	return nil
+}
+
+// setLogWeight tries setting the weight for a single Log of the Log-group.
+// Does not reset the weight and returns error if weight is non-positive and
+// its setting will result innto unability to reach minimal inclusion number.
+func (group *LogGroupInfo) SetLogWeight(logURL string, w float32) error {
+	if !group.LogURLs[logURL] {
+		return fmt.Errorf("trying to assign weight to Log %q not belonging to the group", logURL)
+	}
+	if w < 0.0 {
+		return fmt.Errorf("trying to assign negative weight %v to Log %q", w, logURL)
+	}
+	newWeights := make(map[string]float32)
+	for l, wt := range group.LogWeights {
+		newWeights[l] = wt
+	}
+	newWeights[logURL] = w
+	if !group.satisfyMinimalInclusion(newWeights) {
+		return fmt.Errorf("assigning weight %v to Log %q will result into unability to reach minimal inclusion number %d", w, logURL, group.MinInclusions)
+	}
+	group.wMu.Lock()
+	defer group.wMu.Unlock()
+	group.LogWeights = newWeights
+	return nil
+}
+
+// WeightedLogSample picks a Log from the group-Set and returns its URL.
+// Follows weight distribution LogWeights and ignores Logs specified by
+// ignoreURLs set. If it's not possible (e.g. all Logs are excluded or
+// have 0 weights), returns empty string.
+func (group *LogGroupInfo) WeightedLogSample(ignoreURLs map[string]bool) string {
+	var sum float32
+	for logURL, w := range group.LogWeights {
+		if !ignoreURLs[logURL] {
+			sum += w
+		}
+	}
+	if sum <= 0.0 {
+		return ""
+	}
+
+	r := rand.Float32() * sum
+	for logURL, w := range group.LogWeights {
+		if ignoreURLs[logURL] || w == 0.0 {
```

Comment:
```suggestion
		if ignoreURLs[logURL] {
```

w = 0 isn't doing anything now